### ISSUE TITLE
Add more debug profiling

### ISSUE THREAD

**Issue by jackTabsCode** [matter-ecs/matter#24]
There is a lot of unprofiled or incorrectly profiled code in the library. Profiling all parts of Matter will allow us to narrow down performance issues in Matter's internals.

**Comment by Ukendio** [comment]
I agree with this, but not for the same reason. We want to implement this to allow the users know how much the iteration is taking of their systems' times. 

But if you are just curious about the performance issues in the internals then most of the overhead from the query is clearly the iterator itself, it is very slow at expanding, where unpack is almost 60% of the performance. The rest goes to `next` due to how it works with preventing iterator invalidation. Our overhead comes mostly from supporting ergonomics, and that seems fine with Matter's goals,

**Comment by jackTabsCode** [comment]
I think that, especially with the world rewrite in progress, we should make sure code paths are accurately and extensively profiled. What if we had an API to enable debug profiling? That solves the overhead problem.
